### ISSUE TITLE
Add support for MSIX packages

### ISSUE THREAD

**Issue by qmatteoq** [StefanKert/azuredevops-codesigning-task#8]
Hello,
the current signtool isn't updated to the latest release, included in the Windows 10 1809 SDK. As such, it isn't able to sign MSIX packages, which is the successor of the AppX format.

**Comment by StefanKert** [comment]
Hi @qmatteoq,

thanks so much for sending the PR. I will take a look at it asap!

Thanks

**Comment by StefanKert** [comment]
Do you have a small project that I can use for reproduction? To check if this solves the issue or if it needs additional work?

**Comment by qmatteoq** [comment]
If I publish on GitHub a sample app which outputs a MSIX package, would it be enough? Or is there a way for me to "sideload" the updated Azure DevOps extension before you publish it, so I can test it in my Azure DevOps repo?

**Comment by StefanKert** [comment]
Yeah :) totally! A repo would be fine.In fact, it would also be enough if you hand over the .dll to sign. I think I´ll then add a test cases to make sure that we do not lose support for this in the future :).

Installing the built version of the extension would also be possible :). Whatever you prefer.

**Comment by qmatteoq** [comment]
I've uploaded a folder here https://1drv.ms/f/s!AhGLm1VCMcxP7qIsB2Jhp0WixKfilw which contains a MSIX package and a PFX certificate which can be used to sign it. The certificate's password is "12345". 
If you want to share me a built version of the extension, I'll be also more than happy to test it also in my pipeline!

**Comment by StefanKert** [comment]
@qmatteoq would it be ok for you if I add this stuff to the repository? I would like to have some basic integration tests to let everyone know how this extension works and wanted to double check if it is ok to keep these examples.

**Comment by StefanKert** [comment]
NVM :) i created a demo project. wasn´t that difficult anyways :) support is looking good. It would be amazing if you could give this version a try:

https://dev.azure.com/stefankert/Public/_build/results?buildId=80

**Comment by qmatteoq** [comment]
Thanks for this! I'm travelling today, but I'll make sure to try it during the weekend and let you know 😊

**Comment by StefanKert** [comment]
All good :) thanks for the reply. I will try it as well and as soon as I get your feedback push the stuff to the marketplace. Thanks once again for your help :) greatly appreciated.

**Comment by StefanKert** [comment]
I did some checks in Azure DevOps and everything seems working :) 2.0.0 has been released! https://marketplace.visualstudio.com/items?itemName=stefankert.codesigning

I´ll leave this open until you had time to test it to make sure this issue is resolved.

**Comment by qmatteoq** [comment]
Hello @StefanKert and thanks for your work! Is there anything special I have to do to test the new version?
I switched to use v2 in my pipeline and, in fact, in the logs I can see this:

`
[command]D:\a\_tasks\codesigning_0e0f3bf7-d96c-45d6-aa76-f9afb71fb77e\2.0.0\signtool.exe
`
However, the signing still fails because it complains it can't recognise MSIX:

`
2019-03-30T21:02:47.7077352Z ##[section]Starting: Code Signing AppX
2019-03-30T21:02:47.7185486Z ==============================================================================
2019-03-30T21:02:47.7185963Z Task         : Code Signing
2019-03-30T21:02:47.7186326Z Description  : Codesign assemblies and applications
2019-03-30T21:02:47.7186452Z Version      : 2.0.0
2019-03-30T21:02:47.7186552Z Author       : Stefan Kert
2019-03-30T21:02:47.7186680Z Help         : v2.0.0 - [More Information](https://github.com/StefanKert/azuredevops-codesigning-task#readme)
2019-03-30T21:02:47.7186863Z ==============================================================================
2019-03-30T21:02:47.9545452Z Signing file: D:\a\r1\a\_ContosoExpenses-Basic-Universal Windows Platform-CI\drop\ContosoExpenses.Package_2019.3.17.0_Test\ContosoExpenses.Package_2019.3.17.0_x86.msixbundle
2019-03-30T21:02:47.9582597Z [command]D:\a\_tasks\codesigning_0e0f3bf7-d96c-45d6-aa76-f9afb71fb77e\2.0.0\signtool.exe sign /fd SHA256 /t http://timestamp.digicert.com /f D:\a\_temp\ContosoLab.pfx /p *** "D:\a\r1\a\_ContosoExpenses-Basic-Universal Windows Platform-CI\drop\ContosoExpenses.Package_2019.3.17.0_Test\ContosoExpenses.Package_2019.3.17.0_x86.msixbundle"
2019-03-30T21:03:00.2596426Z Done Adding Additional Store
2019-03-30T21:03:00.3020396Z 
2019-03-30T21:03:00.7555878Z Number of errors: 1
2019-03-30T21:03:00.7557688Z 
2019-03-30T21:03:00.7558148Z SignTool Error: This file format cannot be signed because it is not recognized.
2019-03-30T21:03:00.7558630Z SignTool Error: An error occurred while attempting to sign: D:\a\r1\a\_ContosoExpenses-Basic-Universal Windows Platform-CI\drop\ContosoExpenses.Package_2019.3.17.0_Test\ContosoExpenses.Package_2019.3.17.0_x86.msixbundle
2019-03-30T21:03:00.7558876Z 
2019-03-30T21:03:00.7704636Z ##[error]Error: The process 'D:\a\_tasks\codesigning_0e0f3bf7-d96c-45d6-aa76-f9afb71fb77e\2.0.0\signtool.exe' failed with exit code 1
2019-03-30T21:03:00.7719940Z ##[section]Finishing: Code Signing AppX
`
However, if I try the same command on my machine on the artifact generated by the build pipeline it works without issues.

Thanks!

**Comment by qmatteoq** [comment]
It's really weird because I've cloned the most recent repo on my local machine and I've tried to use the signtool.exe inside the task folder instead of my own one. Also in this case the signing completes just fine. It seems that the problem is specific to the version released on the Azure DevOps marketplace. Unless I'm doing something wrong, which might not be so unlikely :-D

**Comment by qmatteoq** [comment]
Hello @StefanKert , did you have the chance to take a look at this? Thanks!

**Comment by StefanKert** [comment]
Hi @qmatteoq,

sorry for the delay. I did not find the time yet :(, but I will take a look at this tomorrow :)

**Comment by qmatteoq** [comment]
No worries at all! I just wanted to make sure the notification didn't go missing 😊 Thanks for your efforts!

**Comment by StefanKert** [comment]
Good morning @qmatteoq  :),

I´ve just seen that you tried to sign a .msixappbundle. Do you have an example file I can use for the testing this? Because currently I am only testing an appxbundle, which is (at least from what I understand.. I am not very familiar with this stuff.. yet :)) not the same as msixbundle

**Comment by qmatteoq** [comment]
Hello @StefanKert ! I think there was a misunderstanding, let me try to explain the whole scenario. AppX (and appxbundle) were already perfectly supported by your tool. AppX, in fact, is supported since the Windows 8 times, so every version of the signtool package released since then was able to sign AppX packages.

Windows 10 1809 has introduced support for the next generation of AppX, called MSIX. Under the hood it's the same concepts, but it further enhance the packaging story to be more suitable for all the developers and all the IT Pros, regardless of the technology the application is built on top. You can learn more at http://aka.ms/msix

Since MSIX support has been introduced in 1809, you need the signtool version that comes with the Windows 10 1809 SDK to sign a MSIX package. This was the purpose of my PR: I've updated the signtool.exe included in your project with the 1809 one. The PR seems to have been merged just fine. I've tried to clone the project on my computer and, by using the signtool version included in the task folder, I was able to sign a MSIX package just fine. However, the version released on the Marketplace returns me an error that it doesn't recognize the MSIX format, so it seems there's a mismatch.

I've uploaded a folder here https://1drv.ms/f/s!AhGLm1VCMcxP7qIsB2Jhp0WixKfilw which contains a MSIX package and a PFX certificate which can be used to sign it. The password is 12345.

Let me know if I can help further!

**Comment by StefanKert** [comment]
Hi @qmatteoq,

sorry for the missunderstanding. I am not very familiar with desktop development these days, but what you describe sounds pretty cool :) I really need to dig into desktop development again!

I tried to reproduce the things you mentioned and it is actually failing:

https://dev.azure.com/stefankert/Public/_build%2Fresults?buildId=104&_a=summary

So I can confirm that you did nothing wrong :) let me double check if there is something wrong with the current deployment :)! 

Thanks for your help!

**Comment by StefanKert** [comment]
Alright,

It seems like appx files cannot be signed with certificates with a different CN than the creator. I committed the pfx and the appxbundle you sent and started a build and it is working properly:

https://dev.azure.com/stefankert/Public/_build/results?buildId=107&_a=summary (Step Sign appxbundle).

Is it possible that msixbundles and appxbundles behave differntly? Because the file you uploaded is a appxbundle. Do you have a msibxundle as well? Or is there any easy way to create one?

**Comment by qmatteoq** [comment]
Yes, correct. The requirement is that the Publisher declared in the manifest (a special XML file which describes the app and stored inside the AppX / MSIX package) must match with the subject of the certificate.

I apologize for my mistake, now I have uploaded here a proper .msixbundle file with a matching certificate (the password is still 12345). You can get it from here: https://1drv.ms/f/s!AhGLm1VCMcxP7qIsB2Jhp0WixKfilw

I confirm that using the signtool.exe included in the task folder I can sign the package without errors by using the following command:

signtool sign /a /v /fd SHA256 /f "ContosoExpenses.Package_1.0.5.0_x86.cer" /p "12345" ContosoExpenses.Package_1.0.5.0_x86.msixbundle

Let me know if I can help!

**Comment by StefanKert** [comment]
Thanks!

I managed to get a build signing the msix package :) 

https://dev.azure.com/stefankert/Public/_build/results?buildId=120&_a=summary

Are you using hosted agents? Can you double check if it works with the __Hosted Windows 2019 with VS2019__.

I am not 100 percent sure but it seems like the vs2017 images contain the wrong Windows10 SDK or there is any other dependency that is missing.

**Comment by qmatteoq** [comment]
Yes, I'm using the hosted agent with Windows 2019 and VS2019. Is the new version of your extension live? Can I just give it a try on my end?

**Comment by qmatteoq** [comment]
Probably I'm doing something wrong on my end, but I don't have any clue about what 😊

This is the artifact produced by my build:

![image](https://user-images.githubusercontent.com/1230332/55671773-c4668280-5893-11e9-845d-5b1d8ca89a49.png)

And this is, instead, the configuration of the signing task:

![image](https://user-images.githubusercontent.com/1230332/55671779-ce888100-5893-11e9-823b-3171922f9c3c.png)

However, the signing task always fails with the following error:

```
2019-04-06T13:02:41.3253826Z SignTool Error: This file format cannot be signed because it is not recognized.
2019-04-06T13:02:41.3254063Z SignTool Error: An error occurred while attempting to sign: D:\a\r1\a\_ContosoExpenses-Basic-Universal Windows Platform-CI\drop\ContosoExpenses.Package_2019.4.10.0_Test\ContosoExpenses.Package_2019.4.10.0_x86.msixbundle
```

This is the command which is being executed:

```
2019-04-06T13:02:19.4405175Z [command]D:\a\_tasks\codesigning_0e0f3bf7-d96c-45d6-aa76-f9afb71fb77e\2.0.0\signtool.exe sign /fd SHA256 /t http://timestamp.digicert.com /f "D:\a\_temp\Matteo Pagani.pfx" /p *** "D:\a\r1\a\_ContosoExpenses-Basic-Universal Windows Platform-CI\drop\ContosoExpenses.Package_2019.4.10.0_Test\ContosoExpenses.Package_2019.4.10.0_x86.msixbundle"
```

The things I can't explain are:

1. If I execute the same exact command with the same certificate and the output of the build, which I've downloaded from Azure DevOps, everything works fine.
2. It's enough to change the configuration of my project so that it outputs an AppX instead of a MSIX to have the sign task complete without any issue.

Thanks for all your help!

**Comment by StefanKert** [comment]
Hi @qmatteoq ,

once again sorry, but during the week I am usually pretty busy :( not really much time for OS.. sorry.

I tried it once again with the msixbundle you sent and if I use the Windows 2019 and VS2019 agent it works. Can you please double check if you really use  Windows 2019? 

When using VS2017 build agent I get the following error:

```
Done Adding Additional Store

SignTool Error: This file format cannot be signed because it is not recognized.
Number of errors: 1

SignTool Error: An error occurred while attempting to sign: D:\a\1\s/tests/test-files/msixbundle/App.msixbundle
```

Which is exactly what you are facing. But if I try to build with the VS2019 everything works as expected. 


This is the yaml for the my build:

```
pool:
  name: Hosted Windows 2019 with VS2019
steps:
- task: stefankert.codesigning.codesigning-task.codesigning@2
  displayName: 'Code Signing MsixBundle'
  inputs:
    secureFileId: '5af32f57-13d6-49df-9487-3c3ffd50e655'
    signCertPassword: 12345
    files: 'tests/test-files/msixbundle/App.msixbundle'
  continueOnError: true
```

Instead of the name it is also possible to have `vmImage: windows-2019`

**Comment by qmatteoq** [comment]
No worries and thanks for bearing with me!
I found out the problem and that was totally my fault. I had setup VS2019 as hosted agent for the build pipeline (being my app based on .NET Core 3.0), but I left the release pipeline (where the code signing operation is happening) on VS2017. As soon as I have moved it to VS2019, the problem was gone.

Thanks for your help and sorry for this mistake! Feel free to close the issue!
